packages/datadog-plugin-jest/src/index.js

```diff
@@ -0,0 +1,197 @@
+      promiseExec('git rev-parse HEAD')
+    ])
+    return {
+      repository: sanitizeStdout(readRepository),
+      branch: sanitizeStdout(readBranch),
+      commit: sanitizeStdout(readCommit)
+    }
+  } catch (e) {
+    return {
+      repository: '',
+      branch: '',
+      commit: ''
+    }
+  }
+}
+
+function finishStartedSpans () {
+  global.tracer
+    .scope()
+    .active()
+    ._spanContext._trace.started.forEach((span) => {
+      span.finish()
+    })
+}
+
+function setTestStatus (status) {
+  global.tracer.scope().active().setTag(TEST_STATUS, status)
+}
+
+function addTestMetadata ({ testName, testSuite }) {
+  global.tracer
+    .scope()
+    .active()
+    .addTags({
+      [TEST_TYPE]: 'test',
+      [TEST_NAME]: testName,
+      [TEST_SUITE]: testSuite
+    })
+}
+
+module.exports = function (BaseEnvironment) {
+  return class DatadogJestEnvironment extends BaseEnvironment {
+    constructor (config, context) {
+      super(config, context)
+      this.testSuite = context.testPath.replace(`${config.rootDir}/`, '')
+      this.rootDir = config.rootDir
+    }
+    async setup () {
+      if (!global.tracer) {
+        const ciMetadata = getCIMetadata()
+        const { repository, branch, commit } = await getGitInformation()
+        global.tracer = require('../../dd-trace').init({
+          sampleRate: 1,
+          flushInterval: 1,
+          startupLogs: false,
+          ingestion: {
+            sampleRate: 1
+          },
+          tags: {
+            ...ciMetadata,
+            [GIT_COMMIT_SHA]: commit,
+            [GIT_BRANCH]: branch,
+            [GIT_REPOSITORY_URL]: repository,
+            [BUILD_SOURCE_ROOT]: this.rootDir,
+            [TEST_FRAMEWORK]: 'jest'
+          }
+        })
+      }
+
+      return super.setup()
+    }
+    async teardown () {
+      await new Promise((resolve) => {
+        global.tracer._tracer._exporter._writer.flush(resolve)
+      })
+      return super.teardown()
+    }
+
+    async handleTestEvent (event) {
+      if (event.name === 'test_skip' || event.name === 'test_todo') {
+        global.tracer.trace(
+          'jest.test',
+          { type: 'test', resource: `${event.test.parent.name}.${event.test.name}` },
+          (span) => {
+            span.addTags({
+              [TEST_TYPE]: 'test',
+              [TEST_NAME]: event.test.name,
+              [TEST_SUITE]: this.testSuite,
+              [TEST_STATUS]: 'skip'
+            })
+          }
+        )
+      }
+      if (event.name === 'test_start') {
+        const originalSpecFunction = event.test.fn
+        if (originalSpecFunction.length) {
+          event.test.fn = global.tracer.wrap(
+            'jest.test',
+            { type: 'test', resource: `${event.test.parent.name}.${event.test.name}` },
+            () => {
+              addTestMetadata({ testName: event.test.name, testSuite: this.testSuite })
+              return new Promise((resolve, reject) => {
```

Comment:
For callback functions, you can just promisify them and then handle them the same way you handle promises below.